### ISSUE TITLE
Posting to server using safari browser does not work.

### ISSUE THREAD

**Issue by byrning** [photopea/photopea#1670]
Hi,

So it appears that posting data to server when using file > save does not work in safari browser (12.1.2 mojave).  It works fine in chrome and firefox.  What's happening is that the posted content is completely empty.   I think the reason for this is that you're not setting the Content-Type when posting the data and that's causing CORS issues in safari.

Here are the safari headers:

:method: POST
:scheme: https
Accept: */*
Content-Type:  <!-- problem it's empty
Origin: https://www.photopea.com
Content-Length: 0  <!-- problem it's 0
etc

Possibly set the content-type on your end to "multipart/form-data" ?

**Comment by photopea** [comment]
Hi, btw. is Photopea in an iframe in this case?

**Comment by byrning** [comment]
Hi,

Yes it is, I forgot to mention that.   And once again the setup works fine in chrome and firefox, just not safari.

**Comment by photopea** [comment]
Safari seems to restrict websites a lot, which are in iframes on different domains. 

Do you see an error regarding IndexedDB in a web console, when you open Photopea?

    SecurityError:IDBFactory.open() called in an invalid security context

**Comment by byrning** [comment]
Hmm, so no I'm not seeing anything like that.

But I just tried to save a couple times now in safari and the issue is intermittent, it works sometimes and other times not - don't you just love those kinds of bugs/quirks.  The post request always goes through but sometimes the content-length is 0 and posted data empty.  So I hit File > Save and content-length=0, hit File > Save again and then it works.  Right now it seems to be working more often than not. 
 
Weird, I'll reach back out if I can find any additional info. Thanks!

**Comment by photopea** [comment]
Dont close it yet.

Did it happen only for big files, or all files?

**Comment by byrning** [comment]
Well this particular file is only 480x640 and 41kb (see attached).   I open it and image > transform > rotate it and sometimes save works and other times not.  It if just keep clicking file > save it eventually saves.  Maybe there is an issue with safari when you're getting the image and json ready for posting?!?

![img-7117b](https://user-images.githubusercontent.com/9835589/64677762-3a27d900-d446-11e9-9901-00f72e2fe7d3.jpg)

**Comment by photopea** [comment]
Hi, I set the Content-Type of the HTTP request to "application/octet-stream". Does it work better now?

**Comment by byrning** [comment]
Hi, it still happening with completely random results.  The content-type is now set, but sometimes content-length is 0 and the posted json/image is empty.  Repeating file > save results in completely random results.

I did 4 tests with the same image with 10 file > saves and closing and reopening iframe between tests:
800x600 image: 10/10 worked.
600x800 rotated image: 3 worked, 7 failed. 
800x600 image: 1 worked, 9 failed. 
600x800 rotated image: 1 worked, 9 failed. 

I tried a smaller image just to see if there was any difference:
168x250: 10/10 worked.
250x168: 9 worked, 1 failed.
168x250: 10/10 worked.
250x168: 10/10 worked.

So larger images seem to have more issues.

I went into developer settings and turned off safari's cross origin restrictions to rule out CORS being the problem and the issue still happens.  Possibly there is a content-length calculation issue or canvas conversion issue or something like that results in the posted body being completely empty sometimes.  Very strange.

**Comment by photopea** [comment]
If the problem was in CORS, it would either always work, or would never work.

So it relates only to Safari, other browsers always save as expected? I just want to make sure, that the problem is not in your server.

Are you saving to the server as a PSD? When you save it locally, how large is it?

**Comment by byrning** [comment]
- If the problem was in CORS, it would either always work, or would never work.
In theory yes if safari was working 100% in enforcing rules.  I wanted to rule out CORS entirely.

- I just want to make sure, that the problem is not in your server.
My server has no control over the safari request, and the request has a content-length of 0 when the issue happens, so that would be client side issue.

I think I've tracked down the issue though, I think having more than one save format is causing the issue, especially "png".  I had formats:[ "png", "jpg:0.90", "psd:true" ] which is obviously completely overkill.  Removing "png" from that list results in an error once in a while and removing both "png" and "psd:true" seems to resolve the issue completely, so I just have "jpg:0.90".

I then tried just "png" as a format and the save fails often.

**Comment by photopea** [comment]
Can you also answer another question: 

It relates only to Safari, other browsers always save as expected?

**Comment by byrning** [comment]
Correct, firefox and chrome save 100% without error even with 3 save formats on mac and  pc, whereas safari chokes a lot with more than 1 save format and especially "png".  Safari with 1 save format "jpg" seems to be more reliable, like about 80% of saves work.
